### ISSUE TITLE
[FEATURE]: Add skipping logic for failed groups in the “apply_permissions” task of the "migrate-groups-experimental" workflow

### ISSUE THREAD

**Issue by amandeep-db** [databrickslabs/ucx#1914]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Currently the “apply_permissions” task in the "migrate-groups-experimental" workflow fails as soon as we encounter an issue with a group (Shared both the error codes in the additional context). **Can we handle these failure and continue over to the next group while maintaining the list for the failed groups. And persist the failed groups in a new table at the end of the process?** 

**By this we can atleast complete the migration and analyze the problematic groups later. Also, something customer requested**

### Proposed Solution

Skip over to the next group while maintaining the list for the failed groups during the migration. And persist the failed groups in a new table at the end of the process for analysis

### Additional Context

Example error codes:

1. **Preexisting Permissions**:

< 400 Bad Request
< {
<   “error_code”: “BAD_REQUEST”,
<   “message”: “Account group <Group_Name> already has permissions”
< }

2. **Conflicting Permissions**:

< 500 Internal Server Error
< {
<   "error_code": "INTERNAL_ERROR",
<   "message": "Error when updating permissions table"
< }

**Comment by JCZuurmond** [comment]
Just discussed in the stand-up:
1. Create integration test to cover the case
2. Skip the error to unblock

Then fix it better, either by:
1. Logging what permissions were conflicting
2. Or check beforehand what permissions are already there

**Comment by JCZuurmond** [comment]
@amandeep-db : Why did you name it "Conflicting Permissions"? The context of the issue does not say anything about a conflict.

**Comment by asnare** [comment]
The area of code being discussed starts here: https://github.com/databrickslabs/ucx/blob/fb47095a80bd14af875f2a03df2ec8508be3e61e/src/databricks/labs/ucx/workspace_access/manager.py#L46

**Comment by amandeep-db** [comment]
@JCZuurmond thanks for your questions.

Yes, better way is to verify the permissions beforehand. But I don't think we have a way currently to do that as the permission verification only happens during the runtime when the new migration API is called by UCX workflow. 

I am working with the engineer to see if we have a way to verify the permissions beforehand. But are you recommending adding another task in the current workflow for permission validation? Not sure if we have a standalone API for that currently.

Also, we already discussed about the partial migration scenario but problematic groups can be worked with the engineering team for resolution given we will have the complete list for these and something customer is aligned with.


> Why did you name it "Conflicting Permissions"? The context of the issue does not say anything about a conflict.
Yes, the response code is just 5xx but it's the engineer who worked on this API provided this issue description.

**Comment by asnare** [comment]
PR #1924 will partially solve this problem; when merged the group migration will continue even if an error is encountered migrating the permissions for a group. After permissions have been migrated for the rest of the groups the task will still fail. While running the problem groups will appear as errors in the log output.

**Comment by ericvergnaud** [comment]
0.5 pw seems a bit agressive because writing tests for this could be quite demanding

**Comment by JCZuurmond** [comment]
@ericvergnaud : Update the estimate accordingly

**Comment by JCZuurmond** [comment]
@nfx : Could you chime in on the partial group migration as mentioned [here](https://github.com/databrickslabs/ucx/issues/1914#issuecomment-2180568190)?

> Can we handle these failure and continue over to the next group 

Done in #1924, still ending with a failing group migration if one of the groups could not be migrated

> while maintaining the list for the failed groups. And persist the failed groups in a new table at the end of the process?

We can improve the output of the failing group. I have doubts about adding a new table, to the best of my knowledge, we do not keep migration failures in a separate table -> Need to investigate how we handle this with tables.

> Yes, better way is to verify the permissions beforehand. But I don't think we have a way currently to do that as the permission verification only happens during the runtime when the new migration API is called by UCX workflow.

I am afraid that can not be done. It requires a "just do it and ask for permission later" approach.

> Also, we already discussed about the partial migration scenario but problematic groups can be worked with the engineering team for resolution given we will have the complete list for these and something customer is aligned with.

I think this is the main thing that we are lacking at the moment

## Suggestion

1. #1924 avoids failing early when one of the groups in the group migration fails
2. Extend the group migration to provide a list with the groups that can not be migrated
   a. Still let the migration fail, but with the message that it can be resolved manually and a disclaimer that further ucx migrations depend on the group migration to be done completed correctly.
   b. Decide where to persist the list with failed groups.
   c. Make sure that the failed groups list contains all relevant information:
       i. Group names
       ii. Group members
       iii. Permissions
       iv. Admin or not
       v. More?

**Comment by nfx** [comment]
> We can improve the output of the failing group. I have doubts about adding a new table, to the best of my knowledge, we do not keep migration failures in a separate table -> Need to investigate how we handle this with tables.

the simplest would be to consistently prefix the logs with `failed-group-migration: {{display_name}}: ...` and display that on a dashboard sourced from the logs table aka `SELECT * FROM logs WHERE message LIKE 'failed-group-migration: %'`. this is the minimal effort.

**Comment by asnare** [comment]
> […] and display that on a dashboard sourced from the logs […]

Yesterday during discussions we agreed that for group migration we will create a separate dashboard, and the widget displaying failed migrations will be shown there.
